lib/Corres_Cases.thy

```diff
@@ -0,0 +1,325 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Corres_Cases
+imports Corres_UL
+begin
+
+text \<open>
+  This file defines the following main methods for safe data type case distinctions on
+  corres/corres_underlying predicates.
+
+  \<^item> corres_cases_left: case distinction on abstract monad
+  \<^item> corres_cases_right: case distinction on concrete monad
+  \<^item> corres_cases: try corres_cases_left, then corres_cases_right
+  \<^item> corres_cases_both: simultaneous (quadratic) case distinction on both sides, with safe
+                       elimination of trivially contradictory cases.
+
+  The first 3 methods take no arguments, corres_cases_both takes an optional simp argument to,
+  for example, unfold relations that synchronise cases between abstract and concrete side.
+
+  The case distinctions work if the entire monad is a "case" statement, or if the monad is a
+  @{const bind} or @{const bindE} term with a "case" statement in the head position.
+
+  There is an existing method for case distinctions (@{method wpc}), but this method is not
+  flexible enough for @{term corres}: consider the goal
+  @{text "\<And>x. corres r (?G x) ?G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) m"} -- if we perform
+  case distinction on @{term x}, then we can transform @{text "?G x"} into
+  @{text "\<lambda>x s. (x = None \<longrightarrow> ?Q1 x s) \<and> (\<forall>y. x = Some y \<longrightarrow> ?Q2 x y s)"},
+  but we cannot do the same with @{text "?G'"}, because @{text "?G'"} does not depend on @{text x}.
+  The best we can do is @{text "?G' \<leadsto> \<lambda>s. ?A s \<and> ?B s"}, which so far seems to be good enough
+  in our manual proofs.
+
+  The @{method wpc} method will try to treat both preconditions the same and fail on @{text "?G'"}.
+  Extending @{method wpc} so far that it can deal with both situations (uniform treatment of
```

Comment:
Extending X so far that it -> Extending X until it

Reply:
Hm, "until" doesn't sound right to me. Extending `wpc` is not a process that goes on until we reach enough generality. I'm more trying to express that including the feature in `wpc` would stretch the concepts there by quite a bit and make things messy. It'd be completely possible, but it wouldn't be nice.

Reply:
Your version nears German, so if you want that, maybe:
`Overextending X to deal with both situations?`

Reply:
Ok, have reformulated and shrunk that entire paragraph. No more stretching :-)

Reply:
As another alternative, maybe `Extending @{method wpc} enough so that it could deal with both situations`?

Reply:
Sorry, looks like I should have refreshed first, it reads well to me now.
Keeping my pedantic hat on, I think my only other suggestion would be to split the sentence up
```
 even more messy than it already is. We instead re-use the general idea here (in Eisbach instead
  of ML), and leave the @{method wpc} setup clean for other uses.
  ```